contrib/babelfishpg_tsql/src/pl_exec.c

```diff
@@ -10014,3 +10033,62 @@ pltsql_clone_estate_err(PLtsql_estate_err *err)
 	memcpy(clone, err, sizeof(PLtsql_estate_err));
 	return clone;
 }
+
+bool reset_search_path(PLtsql_stmt_execsql *stmt, char *old_search_path)
+{
+	PLExecStateCallStack *top_es_entry;
+	char		*cur_dbname = get_cur_db_name();
+	char 		*new_search_path;
+	char 		*physical_schema;
+	char		*dbo_schema;
+	const char	*user = NULL;
+	const char	*schema;
+	top_es_entry = exec_state_call_stack->next;
+	while(top_es_entry != NULL)
+	{
+		/* traverse through the estate stack. If the occurrence of
+		 * exec in the call stack, update the search path.
+		 */
+		if(top_es_entry->estate && top_es_entry->estate->err_stmt &&
+			top_es_entry->estate->err_stmt->cmd_type == PLTSQL_STMT_EXEC)
+		{
+			if(top_es_entry->estate->schema_name != NULL && stmt->is_dml)
+			{
+				if (top_es_entry->estate->db_name == NULL)
+				{
+					physical_schema = get_physical_schema_name(cur_dbname, top_es_entry->estate->schema_name);
+					dbo_schema = get_dbo_schema_name(cur_dbname);
+				}
+				else
+				{
+					physical_schema = get_physical_schema_name(top_es_entry->estate->db_name, top_es_entry->estate->schema_name);
+					dbo_schema = get_dbo_schema_name(top_es_entry->estate->db_name);
+				}
+				new_search_path = psprintf("%s, %s, %s", physical_schema, dbo_schema, old_search_path);
+				/* Add the schema where the object is referenced and dbo schema to the new search path */
+				(void) set_config_option("search_path", new_search_path,
+								PGC_USERSET, PGC_S_SESSION,
+								GUC_ACTION_SAVE, true, 0, false);
+				return true;
+			}
+			else if(top_es_entry->estate->db_name != NULL && stmt->is_ddl)
+			{
+				user = get_user_for_database(top_es_entry->estate->db_name);
+				schema = get_authid_user_ext_schema_name(top_es_entry->estate->db_name, user);
+				physical_schema = get_physical_schema_name(top_es_entry->estate->db_name, schema);
+				new_search_path = psprintf("%s, %s", physical_schema, old_search_path);
+				/* Add default schema to the new search path */
+				(void) set_config_option("search_path", new_search_path,
+								PGC_USERSET, PGC_S_SESSION,
+								GUC_ACTION_SAVE, true, 0, false);
+				return true;
+			}
+		}
+		/* if the stmt is inside an exec_batch, return false */
+		else if(top_es_entry->estate && top_es_entry->estate->err_stmt &&
+				top_es_entry->estate->err_stmt->cmd_type == PLTSQL_STMT_EXEC_BATCH)
```

Comment:
Because in case the ddl/dml stmt is found inside the exec_batch, we don't need to iterate and check if it is inside an EXEC stmt and go with the default behaviour.